### PR TITLE
Added support for no-resolution images

### PR DESCRIPTION
**Pull request by afiglarz**
Addresses Issue #154 

Added check for 72PPI images, allowing users to process these images without requiring a resample.

### CODE REVIEW

**Comment by afiglarz** [comment]
Removed support for 96 PPI images because biomeval does not support them in the same way as 72 PPI images.

File: NFIQ2/NFIQ2Algorithm/src/tool/nfiq2_ui_refresh.cpp [hunk 1]
```diff
@@ -180,81 +182,132 @@ NFIQ2UI::executeSingle(std::shared_ptr<BE::Image::Image> img,
...
 		}
 	}
-	logger->debugMsg("Successfully passed 500 PPI check. Re-sampled?: " +
+	logger->debugMsg("Successfully passed PPI check. Re-sampled?: " +
 	    std::to_string(resampled));
 
-	// At this point - all images are 500PPI or have been converted to that
-	// resolution. Quantization will happen below if necessary.
+	// At this point - all images are 500PPI, have been converted to that
+	// resolution, or are 72PPI.
```
Comment (gfiumara) [inline NFIQ2/NFIQ2Algorithm/src/tool/nfiq2_ui_refresh.cpp:302]:
```suggestion
	// resolution, or assumed to be that resolution.
```

**Comment by gfiumara** [comment]
* `-F` should treat `defaultDPI` images as `requiredDPI` + should include a note that it was performed in CSV output. Please implement.
 * Ensure that errors for single images are not output in CSV format.

**Review by gfiumara — CHANGES_REQUESTED** [review]
* `-F` should treat `defaultDPI` images as `requiredDPI` + should include a note that it was performed in CSV output. Please implement.
 * Ensure that errors for single images are not output in CSV format.

File: NFIQ2/NFIQ2Algorithm/include/nfiq2/tool/nfiq2_ui_refresh.h [hunk 2]
```diff
@@ -24,6 +27,20 @@
 
 namespace NFIQ2UI {
 
+bool quantizeCheck();
```
Comment (gfiumara) [inline NFIQ2/NFIQ2Algorithm/include/nfiq2/tool/nfiq2_ui_refresh.h:30]:
```suggestion
bool askToQuantize();
```

File: NFIQ2/NFIQ2Algorithm/include/nfiq2/tool/nfiq2_ui_refresh.h [hunk 3]
```diff
@@ -24,6 +27,20 @@
...
+
+bool defaultCheck(const std::string &name, const uint16_t defaultDPI,
+    const uint16_t requiredDPI);
+
+bool resampleCheck(const std::string &name, const uint16_t imageDPI,
```
Comment (gfiumara) [inline NFIQ2/NFIQ2Algorithm/include/nfiq2/tool/nfiq2_ui_refresh.h:35]:
```suggestion
bool askToResample(const std::string &name, const uint16_t imageDPI,
```

File: NFIQ2/NFIQ2Algorithm/include/nfiq2/tool/nfiq2_ui_refresh.h [hunk 4]
```diff
@@ -24,6 +27,20 @@
 
 namespace NFIQ2UI {
 
+bool quantizeCheck();
+
+bool defaultCheck(const std::string &name, const uint16_t defaultDPI,
```
Comment (gfiumara) [inline NFIQ2/NFIQ2Algorithm/include/nfiq2/tool/nfiq2_ui_refresh.h:32]:
```suggestion
bool askToResampleIfNoResolution(const std::string &name, const uint16_t defaultDPI,
```

Hmm, maybe you can think of a better name.

File: NFIQ2/NFIQ2Algorithm/src/tool/nfiq2_ui_refresh.cpp [hunk 5]
```diff
@@ -150,121 +242,157 @@ NFIQ2UI::executeSingle(std::shared_ptr<BE::Image::Image> img,
...
-	if ((resolution.xRes != resolution.yRes) || (imageDPI != 500)) {
-		// Possible re-sampling
-		if (flags.force) {
+	static const uint8_t defaultDPI { 72 };
+	static const uint16_t requiredDPI { 500 };
+
+	const NFIQ2UI::ResampleDims resampleDims { imageHeight, imageWidth,
+		imageDPI, requiredDPI };
+	const NFIQ2UI::ImageProps imageProps { name, fingerPosition, quantized,
+		resampled, singleImage };
```
Comment (gfiumara) [inline NFIQ2/NFIQ2Algorithm/src/tool/nfiq2_ui_refresh.cpp:251]: If you're going to use this struct, use it all the time. You now have `resampled` (non-`const`) and `imageProps.resampled` (`const`), etc. You _will_ screw this up. I'm not even going to check the usage here.

Make this non-const and declare it at the top.

File: NFIQ2/NFIQ2Algorithm/src/tool/nfiq2_ui_refresh.cpp [hunk 6]
```diff
@@ -150,121 +242,157 @@ NFIQ2UI::executeSingle(std::shared_ptr<BE::Image::Image> img,
...
-				    quantized, resampled);
-				return;
-
-			} catch (const NFIR::Miscue &e) {
-				const std::string errStr { e.what() };
-				logger->printError(name, fingerPosition,
-				    "'" + errStr + "'", quantized, resampled);
+				postResample = NFIQ2UI::resampleHelper(
+				    grayscaleRawData, logger, resampleDims,
+				    imageProps);
```
Comment (gfiumara) [inline NFIQ2/NFIQ2Algorithm/src/tool/nfiq2_ui_refresh.cpp:260]: `resampled` here is `true` but `imageProps.resampled` is `false`.

File: NFIQ2/NFIQ2Algorithm/src/tool/nfiq2_ui_refresh.cpp [hunk 7]
```diff
@@ -150,121 +242,157 @@ NFIQ2UI::executeSingle(std::shared_ptr<BE::Image::Image> img,
...
-							static_cast<int>(
-							    imageHeight),
-							static_cast<int>(
-							    imageWidth),
-							CV_8U, grayscaleRawData
+		} else if (flags.force && imageDPI == defaultDPI) {
+			// Don't resample, continue as if it was 500 ppi
+		} else if (interactive && !flags.force) {
+			// Ask to resample
+			if (imageDPI == defaultDPI) {
```
Comment (gfiumara) [inline NFIQ2/NFIQ2Algorithm/src/tool/nfiq2_ui_refresh.cpp:269]: `defaultCheck` probably lets you get rid of this if statement?

**Review by gfiumara — CHANGES_REQUESTED** [review]

File: NFIQ2/NFIQ2Algorithm/src/tool/nfiq2_ui_refresh.cpp [hunk 8]
```diff
@@ -50,6 +50,119 @@ namespace BE = BiometricEvaluation;
...
+}
+
+bool
+NFIQ2UI::askIfDefaultResolution(const std::string &name,
+    const uint16_t defaultDPI, const uint16_t requiredDPI)
+{
+	// 72PPI is default resolution
+	if (defaultDPI != 72) {
+		return false;
+	}
```
Comment (gfiumara) [inline NFIQ2/NFIQ2Algorithm/src/tool/nfiq2_ui_refresh.cpp:70]: If you're passing this is a parameter, and it has to be 72, then don't ask for it as a parameter?

File: NFIQ2/NFIQ2Algorithm/src/tool/nfiq2_ui_refresh.cpp [hunk 9]
```diff
@@ -150,121 +267,135 @@ NFIQ2UI::executeSingle(std::shared_ptr<BE::Image::Image> img,
...
+	if (resolution.xRes != resolution.yRes || imageDPI != requiredDPI) {
+		if (flags.force && imageDPI != defaultDPI) {
+			// resample by force
+			imageProps.resampled = true;
+			try {
+				postResample = NFIQ2UI::resampleAndLogError(
+				    grayscaleRawData, dimensionInfo, imageProps,
+				    logger);
+			} catch (const NFIQ2UI::ResampleError &e) {
+				if (e.errorWasHandled()) {
```
Comment (gfiumara) [inline NFIQ2/NFIQ2Algorithm/src/tool/nfiq2_ui_refresh.cpp:285]: I'd argue _you_ don't have to do this, but it doesn't hurt. Just an FYI. You know your logger isn't `nullptr` (right?).

File: NFIQ2/NFIQ2Algorithm/src/tool/nfiq2_ui_refresh.cpp [hunk 10]
```diff
@@ -150,121 +267,135 @@ NFIQ2UI::executeSingle(std::shared_ptr<BE::Image::Image> img,
...
-						    "'" + errStr + "'",
-						    quantized, resampled);
+						postResample = NFIQ2UI::
+						    resampleAndLogError(
+							grayscaleRawData,
+							dimensionInfo,
+							imageProps, logger);
+					} catch (
+					    const NFIQ2UI::ResampleError &e) {
+						if (e.errorWasHandled()) {
```
Comment (gfiumara) [inline NFIQ2/NFIQ2Algorithm/src/tool/nfiq2_ui_refresh.cpp:320]: Same comment.

File: NFIQ2/NFIQ2Algorithm/src/tool/nfiq2_ui_refresh.cpp [hunk 11]
```diff
@@ -50,6 +50,119 @@ namespace BE = BiometricEvaluation;
...
+		"image to " +
+		std::to_string(requiredDPI) + " PPI?" };
+	return NFIQ2UI::yesOrNo(prompt, false, true, true);
+}
+
+cv::Mat
+NFIQ2UI::resampleAndLogError(BE::Memory::uint8Array &grayscaleRawData,
+    const NFIQ2UI::DimensionInfo &dimensionInfo,
+    const NFIQ2UI::ImageProps &imageProps,
+    std::shared_ptr<NFIQ2UI::Log> logger = nullptr)
```
Comment (gfiumara) [inline NFIQ2/NFIQ2Algorithm/src/tool/nfiq2_ui_refresh.cpp:102]: Did this compile? Default params go in the header.

```suggestion
    std::shared_ptr<NFIQ2UI::Log> logger)
```

File: NFIQ2/NFIQ2Algorithm/src/tool/nfiq2_ui_refresh.cpp [hunk 12]
```diff
@@ -50,6 +50,119 @@ namespace BE = BiometricEvaluation;
...
+{
+	cv::Mat postResample {};
+
+	if (logger != nullptr) {
```
Comment (gfiumara) [inline NFIQ2/NFIQ2Algorithm/src/tool/nfiq2_ui_refresh.cpp:106]: Why duplicate all this code? Check for logger in the exception handlers.

File: NFIQ2/NFIQ2Algorithm/src/tool/nfiq2_ui_refresh.cpp [hunk 13]
```diff
@@ -50,6 +50,99 @@ namespace BE = BiometricEvaluation;
...
+		} else {
+			logger->printError(imageProps.name,
+			    imageProps.fingerPosition, errStr,
+			    imageProps.quantized, imageProps.resampled);
+		}
+		throw NFIQ2UI::ResampleError(errStr, true);
+
+	} catch (const NFIR::Miscue &e) {
+		std::string errStr = "'NFIR resample error: ";
+		errStr = errStr.append(e.what()) + "'";
```
Comment (gfiumara) [inline NFIQ2/NFIQ2Algorithm/src/tool/nfiq2_ui_refresh.cpp:128]: Until update:
```suggestion
	} catch (NFIR::Miscue &e) {
		std::string errStr = "'NFIR resample error: ";
		errStr = errStr.append(e.message()) + "'";
```

**Pull request merged**